src/lib/storage/variable_string_dictionary_segment.cpp

```diff
@@ -0,0 +1,149 @@
+  // For a VariableStringDictionarySegment of the max size Chunk::MAX_SIZE, those two values overlap.
+
+  Assert(_offset_vector->size() < std::numeric_limits<ValueID::base_type>::max(), "Input segment too big");
+}
+
+template <typename T>
+std::shared_ptr<const pmr_vector<char>> VariableStringDictionarySegment<T>::dictionary() const {
+  return _dictionary;
+}
+
+template <typename T>
+std::shared_ptr<VariableStringVector> VariableStringDictionarySegment<T>::variable_string_dictionary() const {
+  return std::make_shared<VariableStringVector>(dictionary(), _offset_vector->size());
+}
+
+template <typename T>
+AllTypeVariant VariableStringDictionarySegment<T>::operator[](const ChunkOffset chunk_offset) const {
+  PerformanceWarning("operator[] used");
+  DebugAssert(chunk_offset != INVALID_CHUNK_OFFSET, "Passed chunk offset must be valid.");
+  // TODO(student): Fix access counter.
+  // access_counter[SegmentAccessCounter::AccessType::Dictionary] += 1;
+  const auto value = get_typed_value(chunk_offset);
+  return value ? value.value() : NULL_VALUE;
+}
+
+template <typename T>
+ChunkOffset VariableStringDictionarySegment<T>::size() const {
+  return static_cast<ChunkOffset>(_attribute_vector->size());
+}
+
+template <typename T>
+std::shared_ptr<AbstractSegment> VariableStringDictionarySegment<T>::copy_using_allocator(
+    const PolymorphicAllocator<size_t>& alloc) const {
+  auto new_attribute_vector = _attribute_vector->copy_using_allocator(alloc);
+  auto new_dictionary = std::make_shared<pmr_vector<char>>(*_dictionary, alloc);
+  auto new_offset = std::make_shared<pmr_vector<uint32_t>>(*_offset_vector, alloc);
+  auto copy = std::make_shared<VariableStringDictionarySegment>(std::move(new_dictionary),
+                                                                std::move(new_attribute_vector), std::move(new_offset));
+  copy->access_counter = access_counter;
+  return copy;
+}
+
+template <typename T>
+size_t VariableStringDictionarySegment<T>::memory_usage(const MemoryUsageCalculationMode /*mode*/) const {
+  using OffsetVectorType = typename std::decay<decltype(*_offset_vector->begin())>::type;
+  return _attribute_vector->data_size() + _dictionary->capacity() +
+         _offset_vector->capacity() * sizeof(OffsetVectorType);
+}
+
+template <typename T>
+std::optional<CompressedVectorType> VariableStringDictionarySegment<T>::compressed_vector_type() const {
+  return _attribute_vector->type();
+}
+
+template <typename T>
+EncodingType VariableStringDictionarySegment<T>::encoding_type() const {
+  return EncodingType::VariableStringDictionary;
+}
+
+template <typename T>
+ValueID VariableStringDictionarySegment<T>::lower_bound(const AllTypeVariant& value) const {
+  DebugAssert(!variant_is_null(value), "Null value passed.");
+  //  access_counter[SegmentAccessCounter::AccessType::Dictionary] +=
+  //      static_cast<uint64_t>(std::ceil(std::log2(_offset_vector->size())));
```

Comment:
Any reason to sometimes comment this access_counter out and sometimes not?

Reply:
Ah, just saw the comment regarding the fix above. Perhaps specify what error you encountered, if you do not plan on fixing it.